README.md
## cloudron-codiad

----------

Built from https://git.cloudron.io/cloudron/lamp-app

## Special Information


- **I am not able to test the security of this app, however, given a strong enough password, things should be fine. In either case, I would NOT INSTALL ON PRODUCTION SYSTEMS**
- After installing, make sure to run the Codiad setup via web browser as well! Once you complete that, add the following line to __/app/data/public/config.php__ if you want Cloudron SSO Authentication: `define("AUTH_PATH", "/app/data/public/ldap.php");`

## Currently Installed Packages / Supported Languages

- Follows Cloudron:base:1.0.0 (See: [Cloudron Baseimage Packages](https://cloudron.io/developer/baseimage/#packages)
- Built off Cloudron's LAMP App

**Supported Langauges**

(If I missed one, let me know by opening an issue or Pull Request.)

- JavaScript
- PHP5/7 (PHP 7.0.3 installed, which is backwards compatible with most PHP5 functions and scripts)
- HTML

## Installing on Cloudron
- Make sure you have [Cloudron CLI installed](https://cloudron.io/developer/cli/) on a Linux computer that is NOT your cloudron server. (See docs for details)
- Then do the following:

```
cloudron login
cloudron install --image mitchellurgero/org.urgero.codiad:latest # Might need to change to whatever the latest build is.
```

## Building from source

- **Building and using Build Service to deploy**

```bash
git clone https://github.com/mitchellurgero/cloudron-codiad
cd cloudron-codiad
cloudron login
cloudron build
# Optional
cloudron install
```


- **Building and using local docker to deploy**

*Note: The below instructions have not been tested, please open an issue if they do not work.*

```bash
git clone https://github.com/mitchellurgero/cloudron-codiad
cd cloudron-codiad

## Change as you need!
docker build -t dockername/projectname:tagname .
docker push dockername/projectname:tagname

## End docker changes
cloudron login
cloudron install --image dockerhuburl/dockername/projectname:tagname
```


## Special Ports

- Port 2222 can be used for SFTP.

## Other Considerations

- Logos, Trademarks, etc are copyright of their respective owners.


CHANGELOG
[0.0.3]
- Added MySQL, PostgreSQL, and MongoDB support to manifest.
- Added Redis Support to manifest.
- Added email support to manifest.

[0.0.2]
- Added LDAP/Cloudron support. (See Post Install Information or README for instructions.)

[0.0.1]
- Initial Release.


test/test.js
#!/usr/bin/env node

'use strict';

require('chromedriver');

var execSync = require('child_process').execSync,
    expect = require('expect.js'),
    fs = require('fs'),
    net = require('net'),
    path = require('path'),
    superagent = require('superagent'),
    util = require('util'),
    webdriver = require('selenium-webdriver');

var by = webdriver.By,
    until = webdriver.until,
    Builder = require('selenium-webdriver').Builder;

process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

if (!process.env.USERNAME || !process.env.PASSWORD) {
    console.log('USERNAME and PASSWORD env vars need to be set');
    process.exit(1);
}

describe('Application life cycle test', function () {
    this.timeout(0);

    var server, browser = new Builder().forBrowser('chrome').build();

    before(function (done) {
        var seleniumJar= require('selenium-server-standalone-jar');
        var SeleniumServer = require('selenium-webdriver/remote').SeleniumServer;
        server = new SeleniumServer(seleniumJar.path, { port: 4444 });
        server.start();

        done();
    });

    after(function (done) {
        browser.quit();
        server.stop();
        done();
    });

    var LOCATION = 'test';
    var TEST_TIMEOUT = 50000;
    var app;

    function waitForElement(elem) {
        return browser.wait(until.elementLocated(elem), TEST_TIMEOUT).then(function () {
            return browser.wait(until.elementIsVisible(browser.findElement(elem)), TEST_TIMEOUT);
        });
    }

    function welcomePage(callback) {
        browser.get('https://' + app.fqdn).then(function () {
            return waitForElement(by.xpath('//*[text()="Cloudron LAMP App"]'));
        }).then(function () {
            return waitForElement(by.xpath('//h1[contains(text(), "7.2.15-0ubuntu0.18.04.1")]'));
        }).then(function () {
            callback();
        });
    }

    function uploadedFileExists(callback) {
        browser.get('https://' + app.fqdn + '/test.php').then(function () {
            return waitForElement(by.xpath('//*[text()="this works"]'));
        }).then(function () {
            return waitForElement(by.xpath('//*[text()="' + app.fqdn + '"]'));
        }).then(function () {
            callback();
        });
    }

    function checkIonCube(callback) {
        browser.get('https://' + app.fqdn).then(function () {
            return waitForElement(by.xpath('//td[contains(text(), "ionCube Loader")]'));
            // return waitForElement(by.xpath('//*[contains(text(), "Intrusion&nbsp;Protection&nsbp;from&nbsp;ioncube24.com")]'));
        }).then(function () {
            callback();
        });
    }

    function checkPhpMyAdmin(callback) {
        superagent.get('https://' + app.fqdn + '/phpmyadmin').end(function (error, result) {
            if (error && !error.response) return callback(error); // network error

            if (result.statusCode !== 401) return callback('Expecting 401 error');

            superagent.get('https://' + app.fqdn + '/phpmyadmin')
                .auth(process.env.USERNAME, process.env.PASSWORD)
                .end(function (error, result) {
                if (error) return callback(error);

                if (result.text.indexOf(`${app.fqdn} / mysql | phpMyAdmin`) === -1) { // in the <title>
                    console.log(result.text);
                    return callback(new Error('could not detect phpmyadmin'));
                }

                callback();
            });
        });
    }

    function checkCron(callback) {
        this.timeout(60000 * 2);

        fs.writeFileSync('/tmp/crontab', '* * * * * echo -n "$MYSQL_HOST" > /app/data/public/cron\n', 'utf8');
        execSync('cloudron push /tmp/crontab /app/data/crontab');
        fs.unlinkSync('/tmp/crontab');

        execSync('cloudron restart --wait');

        console.log('Waiting for crontab to trigger');

        setTimeout(function () {
            superagent.get('https://' + app.fqdn + '/cron').end(function (error, result) {
                if (error && !error.response) return callback(error); // network error

                if (result.statusCode !== 200) return callback('Expecting 200, got ' + result.statusCode);

                if (result.text !== 'mysql') return callback('Unexpected text: ' + result.text);

                callback();
            });
        }, 60 * 1000); // give it a minute to run the crontab
    }

    xit('build app', function () {
        execSync('cloudron build', { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
    });

    it('install app', function () {
        execSync('cloudron install --new --wait --location ' + LOCATION, { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
    });

    it('can get app information', function () {
        var inspect = JSON.parse(execSync('cloudron inspect'));

        app = inspect.apps.filter(function (a) { return a.location === LOCATION; })[0];

        expect(app).to.be.an('object');
    });

    it('can view welcome page', welcomePage);
    it('can access ioncube', checkIonCube);
    it('can upload file with sftp', function () {
        // remove from known hosts in case this test was run on other apps with the same domain already
        // if the tests fail here you want below in ~/.ssh/config
        // Host test.cloudron.xyz
        //     StrictHostKeyChecking no
        //     HashKnownHosts no
        console.log('If this test fails, see the comment above this log message');
        execSync(util.format('sed -i \'/%s/d\' -i ~/.ssh/known_hosts', app.fqdn));
        execSync(util.format('lftp sftp://%s:%s@%s:%s  -e "set sftp:auto-confirm yes; cd public/; put test.php; bye"', process.env.USERNAME, process.env.PASSWORD, app.fqdn, app.portBindings.SFTP_PORT));
    });
    it('can get uploaded file', uploadedFileExists);
    it('can access phpmyadmin', checkPhpMyAdmin);
    it('executes cron tasks', checkCron);

    it('backup app', function () {
        execSync('cloudron backup create --app ' + app.id, { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
    });

    it('restore app', function () {
        execSync('cloudron restore --app ' + app.id, { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
    });

    it('can get uploaded file', uploadedFileExists);

    it('move to different location', function () {
        browser.manage().deleteAllCookies();
        execSync('cloudron configure --wait --location ' + LOCATION + '2 --app ' + app.id, { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
        var inspect = JSON.parse(execSync('cloudron inspect'));
        app = inspect.apps.filter(function (a) { return a.location === LOCATION + '2'; })[0];
        expect(app).to.be.an('object');
    });

    it('can get uploaded file', uploadedFileExists);
    it('can access phpmyadmin', checkPhpMyAdmin);
    it('can access ioncube', checkIonCube);

    // disable SFTP
    it('can disable sftp', function () {
        execSync('cloudron configure --wait -p SFTP_PORT=', { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
    });
    it('(nosftp) can view welcome page', welcomePage);
    it('(nosftp cannot upload file with sftp', function (done) {
        var client = new net.Socket();
        client.setTimeout(10000);

        client.connect(2222, app.fqdn, function() {
            client.destroy();
            done(new Error('Connected'));
        });

        client.on('timeout', function () { client.destroy(); done(); }); // the packet just got dropped (good)

        client.on('error', function (error) {
            client.destroy();
            done(new Error('Should have got timeout but got error:' + error.message));
        });
    });

    it('(nosftp) cannot access phpmyadmin', function (done) {
        superagent.get('https://' + app.fqdn + '/phpmyadmin').end(function (error, result) {
            if (error && !error.response) return done(error); // network error

            if (result.statusCode !== 404) return done('Expecting 404 error');

            done();
        });
    });

    it('uninstall app', function () {
        execSync('cloudron uninstall --app ' + app.id, { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
    });

    // test update
    it('can install app', function () {
        execSync('cloudron install --new --wait --appstore-id lamp.cloudronapp --location ' + LOCATION, { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
        var inspect = JSON.parse(execSync('cloudron inspect'));
        app = inspect.apps.filter(function (a) { return a.location === LOCATION; })[0];
        expect(app).to.be.an('object');
    });
    it('can upload file with sftp', function () {
        // remove from known hosts in case this test was run on other apps with the same domain already
        // if the tests fail here you want to set "HashKnownHosts no" in ~/.ssh/config
        execSync(util.format('sed -i \'/%s/d\' -i ~/.ssh/known_hosts', app.fqdn));
        execSync(util.format('lftp sftp://%s:%s@%s:%s  -e "set sftp:auto-confirm yes; cd public/; put test.php; bye"', process.env.USERNAME, process.env.PASSWORD, app.fqdn, app.portBindings.SFTP_PORT));
    });

    it('can update', function () {
        execSync('cloudron install --wait --app ' + LOCATION, { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
    });
    it('can get uploaded file', uploadedFileExists);
    it('can access phpmyadmin', checkPhpMyAdmin);
    it('can access ioncube', checkIonCube);

    it('uninstall app', function () {
        execSync('cloudron uninstall --app ' + app.id, { cwd: path.resolve(__dirname, '..'), stdio: 'inherit' });
    });
});


ldap.php
<?php

/*
 * Codiad LDAP External Authentication Bridge
 *
 * Copyright (C) 2015 Matt Schultz (Korynkai) & QuantuMatriX Technologies (qmxtech.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

///////////////////
// CONFIGURATION //
///////////////////

// The LDAP connection URI of the server.
	$server = getenv('LDAP_URL');

// The DN to search under on the server.
	$basedn = getenv('LDAP_USERS_BASE_DN');

// Use anonymous bind
//  This does not work by default on Active Directory, however this is the 
//  default method for most servers based on the LDAP standard.
//    Optionally one can bind to a user for search on any LDAP server or
//    enable anonymous binds for search on Active Directory, however this
//    allows for any search option.
	$anonbind = false;
    
// LDAP User for bind (if anonymous bind is set to "false").
	$binddn = getenv('LDAP_BIND_DN');
	$bindpass = getenv('LDAP_BIND_PASSWORD');

// The LDAP search filter. If you aren't sure what this is, the official
// IETF RFC definition (quite technical) is here: 
//	http://tools.ietf.org/search/rfc4515
// and you can find another good (easier to follow) reference here
// (shortened CentOS documentation link):
//	http://goo.gl/FOdGp7
// The default will allow a CN or an email to log in (however, the user 
// environments between the CN and email logins would differ).
// Default is: '(&(objectClass=*)(|(cn=$1)(email=$1)))'.
// A couple alternatives for simple set-ups: 
//	CN only: '(&(objectClass=*)(cn=$1))'
//	email only: '(&(objectClass=*)(email=$1))'
	//$filter = "(&(objectClass=*)(|(cn=$1)(email=$1)))";
        $filter = '(&(objectclass=user)(|(username=1)(mail=1)))';
// Optionally create Codiad user if it doesn't already exist. This can be set
// to 'false' if the administrator would like to manually control access to 
// Codiad from within Codiad itself, rather than let the search filter fully
// dictate user access control. 
// 	Default is 'true'.
	$createuser = true;

// The LDAP protocol version to use. Changing this is probably a very bad 
// idea unless you're absolutely positive you are using some completely 
// non-standard or obsolete (pre-2003) version of LDAP.
// 	Default is '3'. Developer heavily discourages changing this.
	$version = 3;

/////////////////////////////////////////////////////////////////////////////
// DO NOT EDIT ANYTHING UNDER THIS LINE UNLESS YOU KNOW WHAT YOU'RE DOING! //
/////////////////////////////////////////////////////////////////////////////
	//include(__DIR__."/config.php");
	// Ensure we have class.user.php so we may use this class.
	require_once( COMPONENTS . "/user/class.user.php" );
    $preauth = '';
	// Check if our session is not logged in.
	if ( !isset( $_SESSION['user'] ) ) {

		// Check if a username and password were posted.
		if ( isset( $_POST['username'] ) && isset( $_POST['password'] ) ) {

			// Create user object.
			$User = new User();

			// Initialize values of user object.
			$User->username = $_POST['username'];
			$User->password = $_POST['password'];

			// Replace user name token in search filter.
			$tfilter = str_replace( "1", $User->username, $filter );

			// Create LDAP connection socket.
			$socket = ldap_connect( $server );

			// Set initial LDAP values.
			ldap_set_option( $socket, LDAP_OPT_PROTOCOL_VERSION, $version );
			ldap_set_option( $socket, LDAP_OPT_REFERRALS, 0 );
			
			// Check if we are using anonymous bind.
			if ( $anonbind == true ) {
			
				// Set preauth flag to value of socket on anonymous bind.
				$preauth = $socket;
			
			} else {
				
				// Set preauth flag using call to ldap_bind on authenticated bind.
				$preauth = ldap_bind( $socket, $binddn, $bindpass );
			
			}

			// Check if LDAP pre-authentication (or socket creation) was a success.
			if ( $preauth == true ) {

				// Search through basedn based on the filter, and count entries.
				$result = ldap_search( $socket, $basedn, $tfilter );
				$count  = ldap_count_entries( $socket, $result );
					var_error_log($count);
				// Ensure count is definitely equal to 1
				if ( $count === 1 ) {

					// Get the entry from the search result, and bind using its DN.
					$data = ldap_get_entries( $socket, $result );
					$auth = ldap_bind( $socket, $data[0]['dn'], $User->password );
					
					// Check the return value of the bind action.
					if ( $auth === -1 ) {

						// Deny login and send message, An LDAP error occurred.
						die( formatJSEND( "error", "An LDAP error has occurred: " . ldap_error($socket) ) );

					} elseif ( $auth == false ) {

						// Invalid login.
						die( formatJSEND( "error", "Invalid user name or password." ) );

					} elseif ( $auth == true ) {

						// Check if user already exists within users.php.
						if ( $User->CheckDuplicate() ) {

							// Check if we can create a user within users.php.
							if ( $createuser == true ) {

							// Save array back to JSON and set the session username.
							$User->users[] = array( 'username' => $User->username, 'password' => null, 'project' => "" );
							saveJSON( "users.php", $User->users );
							$_SESSION['user'] = $User->username;

							} else {

								// Deny login and send message, the user doesn't exist within users.php.
								die( formatJSEND( "error", "User " . $User->username . " does not exist within Codiad." ) );

							}

						} else {

							// Set the session username.
							$_SESSION['user'] = $User->username;

						}

						// Set the session language, if given, or set it to english as default.
						if ( isset( $_POST['language'] ) ) {

							$_SESSION['lang'] = $_POST['language'];

						} else {

							$_SESSION['lang'] = "en";

						}

						// Set the session theme and project.
						$_SESSION['theme'] = $_POST['theme'];
						$_SESSION['project'] = $_POST['project'];

						// Respond by sending verification tokens on success.
						echo formatJSEND( "success", array( 'username' => $User->username ) );
						header( "Location: " . $_SERVER['PHP_SELF'] . "?action=verify" );

					}

				} elseif ( $count > 1 ) {

					// We returned too many results. Error as such.
					die( formatJSEND( "error", "A server error occurred: LDAP filter result is non-unique. Please ensure this is a unique identifier within its context.
												If the problem persists, please contact the webmaster. If you are the webmaster, please check the LDAP filter used." ) );

				} else {

					// Invalid login.
					//var_error_log($auth);
					die( formatJSEND( "error-ldap", "Incorrect user name or password." ) );

				}

			} else {

				// The server is having issues connecting to the LDAP server. Error as such.
				die( formatJSEND( "error", "An error occurred: Cannot connect to LDAP server. Please contact the webmaster. 
											If you are the webmaster, please contact your LDAP server administrator or check if your LDAP server is running." ) );

			}
		}
	}
function var_error_log( $object=null ){
    ob_start();                    // start buffer capture
    var_dump( $object );           // dump the values
    $contents = ob_get_contents(); // put the buffer into a variable
    ob_end_clean();                // end capture
    error_log( $contents );        // log contents of the result of var_dump( $object )
}
?>



start.sh
#!/bin/bash

set -eu

mkdir -p /app/data/public /run/apache2 /run/proftpd /run/cron /run/app/sessions


# check if any index file exists, if it does not, let's clone Codiad to public!
for f in /app/data/public/index.*; do
    [ -e "$f" ] && echo "Do not override existing index file" || git clone https://github.com/mitchellurgero/Codiad /app/data/public
    break
done

#Copy ldap support into the folder.
cp /app/data/ldap.php /app/data/public/ldap.php

if [ ! -f "/app/data/php.ini" ]; then
    cp /etc/php/7.2/apache2/php.ini.orig /app/data/php.ini
else
    crudini --set /app/data/php.ini Session session.gc_probability 1
    crudini --set /app/data/php.ini Session session.gc_divisor 100
fi

# SFTP_PORT can be unset to disable SFTP
disable_sftp="false"
if [[ -z "${SFTP_PORT:-}" ]]; then
    echo "SSH disabled"
    SFTP_PORT=29418 # arbitrary port to keep sshd happy
    disable_sftp="true"
else
    sed -e "s,##SERVER_NAME,${APP_DOMAIN}," \
        -e "s/##SFTP_PORT/${SFTP_PORT}/" \
        -e "s,##LDAP_URL,${LDAP_URL},g" \
        -e "s/##LDAP_BIND_DN/${LDAP_BIND_DN}/g" \
        -e "s/##LDAP_BIND_PASSWORD/${LDAP_BIND_PASSWORD}/g" \
        -e "s/##LDAP_USERS_BASE_DN/${LDAP_USERS_BASE_DN}/g" \
        -e "s/##LDAP_UID/$(id -u www-data)/g" \
        -e "s/##LDAP_GID/$(id -g www-data)/g" \
        /app/code/proftpd.conf.template > /run/proftpd/proftpd.conf

    if [[ -f /app/data/public/index.php ]]; then
        sed -e "s,^sftp -P.*public/$,sftp -P ${SFTP_PORT} ${APP_DOMAIN}:public/," \
            -i /app/data/public/index.php
    fi
fi

if [[ ! -f "/app/data/sftpd/ssh_host_ed25519_key" ]]; then
    echo "Generating ssh host keys"
    mkdir -p /app/data/sftpd
    ssh-keygen -qt rsa -N '' -f /app/data/sftpd/ssh_host_rsa_key
    ssh-keygen -qt dsa -N '' -f /app/data/sftpd/ssh_host_dsa_key
    ssh-keygen -qt ecdsa -N '' -f /app/data/sftpd/ssh_host_ecdsa_key
    ssh-keygen -qt ed25519 -N '' -f /app/data/sftpd/ssh_host_ed25519_key
else
    echo "Reusing existing host keys"
fi

chmod 0600 /app/data/sftpd/*_key
chmod 0644 /app/data/sftpd/*.pub

## Generate apache config. PMA is disabled based on SFTP config
if [[ "${disable_sftp}" == "true" ]]; then
    echo "PMA disabled"
    sed '/.*PMA BEGIN/,/.*PMA END/d' /app/code/lamp.conf > /run/apache2/lamp.conf
else
    sed -e "s@AuthLDAPURL .*@AuthLDAPURL ${LDAP_URL}/${LDAP_USERS_BASE_DN}?username??(objectclass=user)@" \
        -e "s@AuthLDAPBindDN .*@AuthLDAPBindDN ${LDAP_BIND_DN}@" \
        -e "s@AuthLDAPBindPassword .*@AuthLDAPBindPassword ${LDAP_BIND_PASSWORD}@" \
        /app/code/lamp.conf > /run/apache2/lamp.conf
fi

## hook for custom start script in /app/data/run.sh
if [ -f "/app/data/run.sh" ]; then
    /bin/bash /app/data/run.sh
fi

[[ ! -f /app/data/crontab ]] && cp /app/code/crontab.template /app/data/crontab

## configure in-container Crontab
# http://www.gsp.com/cgi-bin/man.cgi?section=5&topic=crontab
if ! (env; cat /app/data/crontab; echo -e '\nMAILTO=""') | crontab -u www-data -; then
    echo "Error importing crontab. Continuing anyway"
else
    echo "Imported crontab"
fi

chown -R www-data:www-data /app/data /run/apache2 /run/proftpd /run/app

echo "Starting supervisord"
exec /usr/bin/supervisord --configuration /etc/supervisor/supervisord.conf --nodaemon -i Lamp


index.php
<html>
<head>
  <title> Cloudron Codiad app </title>

  <style>

    body {
      width: 50%;
      min-width: 640px;
      margin: auto;
      font-family: Helvetica;
      color: #333;
    }

    pre {
      font-family: monospace;
      background: #333;
      color: white;
      border: none;
      width: 99%;
      padding: 10px;
      text-align: left;
      font-size: 13px;
      border-radius: 5px;
      margin-bottom: 15px;
      box-shadow: 0px 1px 12px rgba(0, 0, 0, 0.176);
    }

    h1 {
      text-align: center;
    }

    .center > table {
      width: 100%;
    }

    .h, .e {
      background-color: white !important;
    }

  </style>

</head>
<body>

<h1>Cloudron LAMP App</h1>

<br/>

<h2>MySQL credentials</h2>
<p>Use the following environment variables in the PHP code to access MySQL:</p>
<pre>
getenv("MYSQL_HOST")
getenv("MYSQL_PORT")
getenv("MYSQL_USERNAME")
getenv("MYSQL_PASSWORD")
getenv("MYSQL_DATABASE")
</pre>

<br/>

<h2>SFTP Transfer</h2>
<p>
  You can SFTP files to the <b>public</b> folder using  <a href="https://cyberduck.io/" target="_blank">Cyberduck</a>,
  <a href="https://filezilla-project.org/" target="_blank">FileZilla</a> or <a href="https://www.gftp.org/" target="_blank">gFTP</a>
  (use your cloudron credentials to authenticate).
</p>
<p>
  We recommend disabling SFTP access (from Cloudron's configuration panel) once you have uploaded your code.
</p>
<pre>
sftp -P 2222 surfer.nebulon.info:public/
</pre>

<br/>

<h2>phpMyAdmin access</h2>
<p>
  You can access phpMyAdmin using your Cloudron credentials <a href="/phpmyadmin" target="_blank">here</a>.
</p>
<p>
  Note that disabling SFTP access also disables phpMyAdmin.
</p>

<br/>

<h2>Cron</h2>
<p>
  Just put a file called crontab to the directory /app/data and it will picked up at next app restart. It has to be in the cron syntax without username
  and must end with a newline. For example, the following crontab updates feeds every hour:
</p>
<pre>
0 * * * * php /app/code/update.php --feeds

</pre>
<p>
  Commands are executed as the user www-data. See <a href="https://linux.die.net/man/5/crontab">manpage</a> for a description of the file.
</p>

<br/>

<h2>Addons</h2>
<p>The app is configured to have access to the following Cloudron addons:</p>
<ul>
  <li><a href="https://cloudron.io/developer/addons/#mysql" target="_blank">mysql</a></li>
  <li><a href="https://cloudron.io/developer/addons/#localstorage" target="_blank">localstorage</a></li>
  <li><a href="https://cloudron.io/developer/addons/#sendmail" target="_blank">sendmail</a></li>
  <li><a href="https://cloudron.io/developer/addons/#redis" target="_blank">redis</a></li>
  <li><a href="https://cloudron.io/developer/addons/#ldap" target="_blank">ldap</a></li>
</ul>
<p>Read more about Cloudron addons and how to use them <a href="https://cloudron.io/developer/addons/" target="_blank">here</a>.</p>

<br/>

<h2>PHP Setup</h2>
<?php

echo phpInfo();

?>

</body>
</html>


ftp.js
var ftpd = require('ftpd'),
    fs = require('fs'),
    superagent = require('superagent'),
    path = require('path');

var simpleAuth = process.env.SIMPLE_AUTH_URL && process.env.SIMPLE_AUTH_CLIENT_ID && process.env.API_ORIGIN;

function verifyUser(username, password, callback) {
    if (!simpleAuth) {
        if (username === 'test' && password === 'test') return callback(null);
        else return callback(new Error('auth failed'));
    }

    var authPayload = {
        clientId: process.env.SIMPLE_AUTH_CLIENT_ID,
        username: username,
        password: password
    };

    superagent.post(process.env.SIMPLE_AUTH_URL + '/api/v1/login').send(authPayload).end(function (error, result) {
        if (error && error.status === 401) return callback(new Error('auth failed'));
        if (error) return callback(wrapRestError(error));
        if (result.status !== 200) return callback(new Error('auth failed'));

        callback(null);
    });
}

var server;
var options = {
    host: '0.0.0.0',
    port: process.env.FTP_PORT || 7002,
    tls: null,
};

server = new ftpd.FtpServer(options.host, {
    getInitialCwd: function() {
      return '/';
    },
    getRoot: function() {
      return '/app/data/public';
    },
    pasvPortRangeStart: process.env.FTP_PORT_PASV_0 || 7003,
    pasvPortRangeEnd: process.env.FTP_PORT_PASV_3 || 7006,
    tlsOptions: options.tls,
    allowUnauthorizedTls: true,
    useWriteFile: false,
    useReadFile: false
});

server.on('error', function(error) {
    console.log('FTP Server error:', error);
});

server.on('client:connected', function(connection) {
    var username = null;
    console.log('client connected: ' + connection.remoteAddress);
    connection.on('command:user', function(user, success, failure) {
        if (user) {
            username = user;
            success();
        } else {
            failure();
        }
    });

    connection.on('command:pass', function(password, success, failure) {
        if (!password) return failure();

        verifyUser(username, password, function (error) {
            if (error) failure();
            else success(username);
        });
    });
});

server.debugging = 4;
server.listen(options.port);
console.log('Listening on port ' + options.port);
